File: app/src/main/java/za/org/grassroot2/presenter/BasePresenter.java
package za.org.grassroot2.presenter;

import android.util.Log;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import retrofit2.Response;
import timber.log.Timber;
import za.org.grassroot2.model.exception.AuthenticationInvalidException;
import za.org.grassroot2.model.exception.ServerUnreachableException;
import za.org.grassroot2.view.GrassrootView;


public class BasePresenter<T extends GrassrootView> implements GrassrootPresenter {

    protected T view;
    private CompositeDisposable disposables = new CompositeDisposable();


    public void attach(T view) {
        this.view = view;
    }

    public void detach() {
        disposables.clear();
        this.view = null;
    }

    protected void disposableOnDetach(Disposable d) {
        if (!disposables.isDisposed()) {
            disposables.add(d);
        }
    }


    @Override
    public void handleResponseError(Response response) {
    }

    @Override
    public void handleNetworkConnectionError(Throwable t) {
        Timber.d(t);
        view.closeProgressBar();
        view.handleNoConnection();
    }

    @Override
    public void handleNetworkUploadError(Throwable t) {
        Timber.d(t);
        view.closeProgressBar();
        view.handleNoConnectionUpload();
    }

    public void handleAuthenticationError(AuthenticationInvalidException t) {
        Timber.d(t);
        view.closeProgressBar();
    }

    public void handleServerUnreachableError(ServerUnreachableException e) {
        Timber.d(e);
        view.closeProgressBar();
        view.closeKeyboard();
    }

    // these two should only be called internally, to enforce design
    void handleGenericKnownException(Throwable t) {
        Log.e("ERROR", t.toString());
    }


}

File: app/src/main/java/za/org/grassroot2/presenter/BasePresenter.kt
package za.org.grassroot2.presenter

import android.util.Log

import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Response
import timber.log.Timber
import za.org.grassroot2.model.exception.AuthenticationInvalidException
import za.org.grassroot2.model.exception.ServerUnreachableException
import za.org.grassroot2.view.GrassrootView


open class BasePresenter<T : GrassrootView> : GrassrootPresenter {

    protected lateinit var view: T
    private val disposables = CompositeDisposable()


    fun attach(view: T) {
        this.view = view
    }

    fun detach() {
        disposables.clear()
    }

    protected fun disposableOnDetach(d: Disposable) {
        if (!disposables.isDisposed) {
            disposables.add(d)
        }
    }


    override fun handleResponseError(response: Response<*>) {}

    override fun handleNetworkConnectionError(t: Throwable) {
        Timber.d(t)
        view.closeProgressBar()
        view.handleNoConnection()
    }

    override fun handleNetworkUploadError(t: Throwable) {
        Timber.d(t)
        view.closeProgressBar()
        view.handleNoConnectionUpload()
    }

    override fun handleAuthenticationError(t: AuthenticationInvalidException) {
        Timber.d(t)
        view.closeProgressBar()
    }

    override fun handleServerUnreachableError(e: ServerUnreachableException) {
        Timber.d(e)
        view.closeProgressBar()
        view.closeKeyboard()
    }

    fun io(): Scheduler = Schedulers.io()

    fun main(): Scheduler = AndroidSchedulers.mainThread()

    // these two should only be called internally, to enforce design
    internal fun handleGenericKnownException(t: Throwable) {
        Log.e("ERROR", t.toString())
    }

}
